src/read.cpp

```diff
@@ -38,15 +39,27 @@ CharacterVector read_lines_(List sourceSpec, List locale_, std::vector<std::stri
   LocaleInfo locale(locale_);
   Progress progressBar;
 
+  bool chunked = chunk_size != -1;
+
   R_len_t n = (n_max < 0) ? 10000 : n_max;
+
   CharacterVector out(n);
 
   R_len_t i = 0;
+  R_len_t pos = i + 1;
   for (Token t = tokenizer.nextToken(); t.type() != TOKEN_EOF; t = tokenizer.nextToken()) {
     if (progress && (i + 1) % 25000 == 0)
       progressBar.show(tokenizer.progress());
 
-    if (i >= n) {
+    if (chunked && i >= chunk_size) {
+      if (!as<Function>(callback["continue"])()) {
+        break;
+      }
+      out = Rf_xlengthgets(out, i);
+      as<Function>(callback["receive"])(out, pos);
+      pos = i + 1;
+      i = 0;
+    } else if (!chunked && i >= n) {
       if (n_max < 0) {
         // Estimate rows in full dataset
         n = (i / tokenizer.progress().first) * 1.2;
```

Comment:
That is a good point actually and looks like a bug in the current implementation to me. `out[i++]` seems like the right fix. I am not sure if this is actually a problem in practice, it doesn't look like`TokenizerLine()` can actually produce a `TOKEN_EMPTY` value. Regardless this could easily change in the future, so it should be changed.